### ISSUE TITLE
Question about kitti_eval

### ISSUE THREAD

**Issue by cxzhou95** [HKUST-Aerial-Robotics/Stereo-RCNN#19]
There is 7481 images(with labels) in KITTI dataset. I filtered them to just keep images with cars.
Then I split remaining images into training set and validation set with the same amount (3410).
I train stereo rcnn with train.txt and test it with val.txt. But there is only 3342 txts in dir 'result'

I think it is because the detector detects nothing in some images, right? So I create empty txts in dir 'result' for these images.
 
In my experiment, I want to compare my method with yours in 2D AP and disparity error. I get the box in left img and right img from the 3D box. But, I test 2D AP(IoU=0.7) on left img with kitti 2d object detection eval tool, and it is 86.61,80.39,68.10 which is lower than 98.73 88.48 71.26 in your paper.

And I find that the eval tools you use only evaluates samples whose result files exist at https://github.com/prclibo/kitti_eval.

**Comment by PeiliangLi** [comment]
Hi, complementing empty files won't affect the performance as long as the detector has a low FN rate wrt the KITTI evaluating targets (Min height > 25 pixels, truncation < 0.5, etc).

**Comment by cxzhou95** [comment]
Sorry to bother you again.

To figure this question, I download the pretrained model 'stereo_rcnn_20_6477' and use it to test images in val.txt. Then I use the tool https://github.com/prclibo/kitti_eval (groundtruth dir is the label of KITTI 3d)to get AP:

car_detection AP: 98.640228 88.898941 71.541313
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_orientation.txt
car_orientation AP: 98.115669 87.833221 70.612625
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_orientation.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_detection_ground.txt
car_detection_ground AP: 70.340759 48.950481 41.507397
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection_ground.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_detection_3d.txt
car_detection_3d AP: 55.301403 36.549541 31.224287
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection_3d.pdf'.
Your evaluation results are available at:
/media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result/

It is quite close to the results in Table 1 and Table 2 of Stereo_RCNN.
But I find that the training of Stereo_RCNN uses the label of   ['Car', 'Van' ,'Truck', 'Misc'], but when testing, the eval tool only consider 'Car' in  the labels of KITTI 3d instead of considering all labels with ['Car', 'Van' ,'Truck', 'Misc']. So I rename  ['Van' ,'Truck', 'Misc'] with 'Car' and use the tool test again:

car_detection AP: 89.552559 79.790436 62.627747
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_orientation.txt
car_orientation AP: 89.073715 78.777321 61.815041
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_orientation.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_detection_ground.txt
car_detection_ground AP: 62.577099 42.690716 35.933727
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection_ground.pdf'.
save /media/yazhou/data_drive2/zcx/Stereo-RCNN/models_stereo/result//plot/car_detection_3d.txt
car_detection_3d AP: 48.883743 32.102699 26.643631
PDFCROP 1.38, 2012/11/02 - Copyright (c) 2002-2012 by Heiko Oberdiek.
==> 1 page written on `car_detection_3d.pdf'.

I am not familiar with KITTI 3d detection. So is it a common way of training with ['Car', 'Van' ,'Truck', 'Misc'] and testing with ['Car']? Or just a carelessness?

**Comment by PeiliangLi** [comment]
We only use Car for training. The 'Van' ,'Truck', 'Misc' are collected for deriving the occlusion relations, e.g., some cars are is occluded by Vans, we calculate the boundary keypoints according to the depth of objects. Once we get the boundary keypoints, we only keep the Car for training. See: https://github.com/HKUST-Aerial-Robotics/Stereo-RCNN/blob/master/lib/datasets/kitti.py#L217

**Comment by cxzhou95** [comment]
> We only use Car for training. The 'Van' ,'Truck', 'Misc' are collected for deriving the occlusion relations, e.g., some cars are is occluded by Vans, we calculate the boundary keypoints according to the depth of objects. Once we get the boundary keypoints, we only keep the Car for training. See: https://github.com/HKUST-Aerial-Robotics/Stereo-RCNN/blob/master/lib/datasets/kitti.py#L217

Thanks for explaining. I get it now !

BTW, could you provide the code for calculating stereo AP in Table1?

**Comment by PeiliangLi** [comment]
You can directly modify the code in https://github.com/prclibo/kitti_eval. It doesn't require much effort as we describe in the paper.

**Comment by cxzhou95** [comment]
Okay, I will try it. Thank you!
